include/gridtools/fn/backend2/gpu.hpp

```diff
@@ -0,0 +1,159 @@
+        using block_size_at_dim = meta::second<meta::mp_find<BlockSizes, Dim>>;
+
+        template <class BlockSizes, class Sizes>
+        GT_FUNCTION_DEVICE auto global_thread_index() {
+            using keys_t = get_keys<Sizes>;
+            using ndims_t = meta::length<keys_t>;
+            static_assert(ndims_t::value > 0);
+            if constexpr (ndims_t::value == 1) {
+                using block_dim_x = block_size_at_dim<BlockSizes, meta::at_c<keys_t, 0>>;
+                using values_t = typename keys_t::template values<int>;
+                return values_t(blockIdx.x * block_dim_x::value + threadIdx.x);
+            } else if constexpr (ndims_t::value == 2) {
+                using block_dim_x = block_size_at_dim<BlockSizes, meta::at_c<keys_t, 0>>;
+                using block_dim_y = block_size_at_dim<BlockSizes, meta::at_c<keys_t, 1>>;
+                using values_t = typename keys_t::template values<int, int>;
+                return values_t(
+                    blockIdx.x * block_dim_x::value + threadIdx.x, blockIdx.y * block_dim_y::value + threadIdx.y);
+            } else {
+                using block_dim_x = block_size_at_dim<BlockSizes, meta::at_c<keys_t, 0>>;
+                using block_dim_y = block_size_at_dim<BlockSizes, meta::at_c<keys_t, 1>>;
+                using block_dim_z = block_size_at_dim<BlockSizes, meta::at_c<keys_t, 2>>;
+                using values_t = typename keys_t::template values<int, int, int>;
+                return values_t(blockIdx.x * block_dim_x::value + threadIdx.x,
+                    blockIdx.y * block_dim_y::value + threadIdx.y,
+                    blockIdx.z * block_dim_z::value + threadIdx.z);
+            }
+        }
+
+        template <class Index, class Sizes>
+        GT_FUNCTION_DEVICE bool in_domain(Index const &index, Sizes const &sizes) {
+            auto in_bounds = tuple_util::device::transform(std::less(), index, sizes);
+            return tuple_util::device::fold(std::logical_and(), wstd::move(in_bounds));
+        }
+
+        template <class BlockSizes, class Sizes, class PtrHolder, class Strides, class Fun>
+        __global__ void kernel(Sizes sizes, PtrHolder ptr_holder, Strides strides, Fun fun) {
+            auto thread_idx = global_thread_index<BlockSizes, Sizes>();
+            if (!in_domain(thread_idx, sizes))
+                return;
+            static_assert(meta::length<Sizes>::value <= 3, "higher dimensional computations not implemented");
+            auto ptr = ptr_holder();
+            sid::multi_shift(ptr, strides, thread_idx);
+            fun(ptr, strides);
+        }
+
+        template <class BlockSizes, class Sizes>
+        std::tuple<dim3, dim3> blocks_and_threads(Sizes const &sizes) {
+            using keys_t = get_keys<Sizes>;
+            using ndims_t = meta::length<keys_t>;
+            dim3 blocks(1, 1, 1);
+            dim3 threads(1, 1, 1);
+            threads.x = block_size_at_dim<BlockSizes, meta::at_c<keys_t, 0>>::value;
```

Comment:
I realized that we actually need to schedule one block and one thread in the zero-d case as it could used when applying a column_stage to a 1D storage.